File: laravel-elasticemail-examples/app/Http/Controllers/InboundController.php
<?php

namespace App\Http\Controllers;

use ElasticEmail\Model\MessageAttachment;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;

/**
 * Inbound email pushed by an inbound route with ActionType "NotifyViaHttp".
 *
 * Setup: point the domain's MX record at mx.inbound.elasticemail.com and create a route
 * whose HttpAddress is PUBLIC_URL/api/inbound?token=ELASTICEMAIL_WEBHOOK_TOKEN.
 * Elastic Email POSTs the parsed message as form fields: from_email, from_name, env_from,
 * env_to_list, to_list, header_list, subject, body_text, body_html,
 * att1_name/att1_content (base64), att2_name/att2_content, ...
 */
class InboundController extends Controller
{
    public function handle(Request $request)
    {
        if (!$this->ee->tokenOk($request->query('token'))) {
            return response()->json(['error' => 'Invalid token'], 401);
        }

        $mail = $request->post();
        $attachments = [];
        foreach ($mail as $key => $value) {
            if (preg_match('/^att\d+_name$/', $key)) {
                $content = $mail[str_replace('_name', '_content', $key)] ?? null;
                if ($content) {
                    $attachments[] = new MessageAttachment(['name' => $value, 'binary_content' => $content]);
                }
            }
        }

        Log::info('Inbound email received', [
            'from' => $this->sanitize($mail['from_email'] ?? ''),
            'subject' => $this->sanitize($mail['subject'] ?? ''),
            'attachments' => count($attachments),
        ]);

        $body = $mail['body_html'] ?? '';
        if ($body === '') {
            $body = '<pre>' . e($mail['body_text'] ?? '') . '</pre>';
        }

        $forwardTo = config('elasticemail.contact_email') ?: $this->ee->from();

        try {
            $result = $this->ee->sendHtml(
                $forwardTo,
                'Fwd: ' . ($mail['subject'] ?? '(no subject)'),
                view('emails.inbound-forwarded', [
                    'originalFrom' => $mail['from_email'] ?? '',
                    'originalTo' => $mail['to_list'] ?? '',
                    'originalSubject' => $mail['subject'] ?? '',
                    'originalDate' => now()->toRfc2822String(),
                    'body' => $body,
                ])->render(),
                [
                    'reply_to' => $mail['from_email'] ?? null,
                    'attachments' => $attachments,
                ]
            );

            return response()->json(['received' => true, 'forwardedTo' => $forwardTo, 'forwardedMessageId' => $result->getMessageId()]);
        } catch (\Exception $e) {
            Log::error('Inbound forward failed', ['error' => $e->getMessage()]);
            return $this->fail($e);
        }
    }
}

File: elixir-elasticemail-examples/phoenix_app/lib/phoenix_app_web/controllers/inbound_controller.ex
defmodule PhoenixAppWeb.InboundController do
  use Phoenix.Controller, formats: [:json]
  import PhoenixAppWeb.Helpers
  require Logger

  # Inbound email pushed by an inbound route with ActionType "NotifyViaHttp".
  # Form fields: from_email, from_name, env_from, env_to_list, to_list, header_list,
  # subject, body_text, body_html, att1_name/att1_content (base64), att2_name/att2_content, ...
  def create(conn, params) do
    if token_ok?(params["token"]) do
      forward(conn, params)
    else
      unauthorized(conn)
    end
  end

  defp forward(conn, mail) do
    attachments =
      mail
      |> Map.keys()
      |> Enum.filter(&Regex.match?(~r/^att\d+_name$/, &1))
      |> Enum.map(fn key ->
        %{name: mail[key], content: mail[String.replace(key, "_name", "_content")]}
      end)

    Logger.info(
      "Inbound email from: #{sanitize(mail["from_email"])} subject: #{sanitize(mail["subject"])}"
    )

    names = attachments |> Enum.map(& &1.name) |> Enum.join(", ")
    Logger.info("Attachments: #{if names == "", do: "none", else: names}")

    body_html =
      case mail["body_html"] do
        html when is_binary(html) and html != "" -> html
        _ -> "<pre>#{String.replace(mail["body_text"] || "", "<", "&lt;")}</pre>"
      end

    # Forward a copy to the team inbox
    case ElasticEmail.send_transactional(%{
           "Recipients" => %{"To" => [ElasticEmail.Env.contact_email()]},
           "Content" => %{
             "From" => ElasticEmail.Env.from(),
             "ReplyTo" => mail["from_email"],
             "Subject" => "Fwd: #{mail["subject"] || "(no subject)"}",
             "Body" => [%{"ContentType" => "HTML", "Content" => body_html}],
             "Attachments" =>
               attachments
               |> Enum.filter(&(&1.content not in [nil, ""]))
               |> Enum.map(&%{"Name" => &1.name, "BinaryContent" => &1.content})
           }
         }) do
      {:ok, data} -> json(conn, %{received: true, forwardedMessageId: data["MessageID"]})
      error -> api_error(conn, error)
    end
  end
end
